### PR TITLE
Use posix_memalign/memalign/_aligned_malloc

### PR DESCRIPTION
**Pull request by steven-johnson**
This changes posix_allocator to use posix_memalign(), and adds android_allocator (using memalign) and windows_allocator (using _aligned_malloc/_aligned_free), rather than using the over-allocation trick we have been using.

This will need testing on Android devices before landing.

Note that I didn't attempt to modify qurt_allocator, as I don't know which API  variants are available there; someone from Qualcomm should probably take a look at possible improvements there.

### CODE REVIEW

**Comment by abadams** [comment]
Considerable effort has gone into optimizing malloc for lots of tiny allocations, I imagine. I wonder if posix_memalign is similarly good. This probably needs benchmarking on some schedules with fine-grained compute_ats that use the heap.

**Comment by dsharletg** [comment]
Hexagon does have memalign or posix_memalign, however, it had a wacky bug at one point (I think maybe the alignment and size arguments were swapped in meaning?). This may be fixed/workaround-able. However, it may be the case that if this was a bug at one point, there is no consistent API we can use unless we give up supporting devices older than a certain point (maybe reasonable). @pranavb-ca

File: src/runtime/posix_allocator.cpp [hunk 1]
```diff
@@ -3,25 +3,27 @@
...
     const size_t alignment = halide_malloc_alignment();
-    void *orig = malloc(x + alignment);
-    if (orig == NULL) {
-        // Will result in a failed assertion and a call to halide_error
-        return NULL;
+#ifdef DEBUG_RUNTIME
+    // posix_memalign requires that the alignment be at least sizeof(void*).
+    // Halide should always handle this, but check in Debug mode, just in case.
+    if (alignment < sizeof(void*)) {
+        halide_error(user_context, "halide_default_malloc: alignment is too small\n");
```
Comment (dsharletg) [inline src/runtime/posix_allocator.cpp]: return NULL here(?)

Comment (steven-johnson) [reply src/runtime/posix_allocator.cpp]: Nice catch!

**Comment by steven-johnson** [comment]
(Circling back to this after a couple of weeks away...) 

Adding @pranavb-ca as reviewer to confirm (or not) the issue that @dsharletg mentioned above.

> This probably needs benchmarking on some schedules with fine-grained compute_ats that use the heap.

I'll see what I can find inside Google that fits this requirement.

**Comment by steven-johnson** [comment]
(BTW: this should ideally also have qurt_allocator updated to use posix_memalign before it lands, but I'm not going to undertake that until I hear from someone at Qualcomm re: the bug history of it, per Dillon's comment about whether there is a consistent API we can use.)

**Comment by dsharletg** [comment]
Maybe we should just go ahead with this without doing Hexagon for now? I think this is a good change to make.

@pranavb-ca for any last comments

**Comment by steven-johnson** [comment]
Still waiting for comments/approval from @pranavb-ca

**Comment by dsharletg** [comment]
I was poking around the Hexagon SDK to see if I could find anything. The declaration is just `memalign(size_t, size_t)` so there's no way to tell if the old bug still persists (https://github.com/halide/Halide/pull/2162).

However, while looking, I also found `aligned_alloc`, and this is apparently a C11 standard. If that is widely supported, that might be a good option on all platforms (except maybe windows).

**Comment by steven-johnson** [comment]
`aligned_alloc` is in C11 but (IIRC) not yet in C++. (Thought maybe it's in C++17?)

**Comment by dsharletg** [comment]
Does that matter though? We're extern declaring this stuff anyways, it doesn't matter where the headers come from...

**Comment by steven-johnson** [comment]
If we're building in C++ (rather than C) then we have the C++ runtime, so it matters if the extern-declared function exists or not

**Comment by dsharletg** [comment]
I was under the impression that on most/all platforms, C++ code links to libc (in addition to libc++ in C++ programs)? So much code includes C headers, not C++ headers, and that stuff all works (math.h, stdlib.h, etc.). In fact, I'm pretty sure 'aligned_alloc' comes from stdlib.h.

**Comment by dsharletg** [comment]
Also, isn't C the language that we minimally support at runtime? We just have some C++ sugar on top of a C API (`#ifdef`ed on C++ support).

**Comment by steven-johnson** [comment]
Still waiting for comments/approval from @pranavb-ca

**Comment by pranavb-ca** [comment]
Sorry I completely missed this. Will get back shortly.

**Comment by steven-johnson** [comment]
not super-urgent, but would like to land when we're ready.

**Comment by zvookin** [comment]
The issue re: "don't we just rely on C" is the distinction between "C" and "C11." System libc is not necessarily the same as what a given installed compiler links with. As such, there may be a very long lag for newer C library support on various operating systems. It is possible this is supported everywhere and it will work, but we'd have to check that and I won't be surprised if it doesn't work out. Largely what we depend on when we depend on C is something old enough to have made it to every corner of the world and we try not to force people to e.g. upgrade their OS to run Halide output.

Contrast this to C++ requirements to run the compiler itself, which some consider too cutting edge in requiring a version of C++ less than a decade old :-)

**Comment by zvookin** [comment]
Oh and I should also add that if there is significant win to be had, I'm willing to push on e.g. upgrading libc or to force installing glibc in a standard place on UNIX systems, etc. For this change, I've never been clear on the advantage other than code cleanliness. That's a nice to have, but it isn't worth causing end user hassle for.

**Comment by steven-johnson** [comment]
Another ping to @pranavb-ca.

**Comment by pranavb-ca** [comment]
Hi @steven-johnson, I am back from vacation. I asked @aankit-ca  to take a look at this yesterday and from initial testing it doesn't look like he had any trouble switching to "qurt_memalign". Both of us will do an internal review and get back to you later today/early tomorrow (@aankit-ca is in India)

**Comment by steven-johnson** [comment]
Thanks for the update!

**Comment by aankit-ca** [comment]
We have done some preliminary testing and memalign seems work fine on device and simulator. Just testing on a few more cases. Should the patch for hexagon be a part of this PR or a separate one?

**Comment by steven-johnson** [comment]
Add to this PR I think is best. Feel free to add commits to this branch as you see fit.

**Comment by aankit-ca** [comment]
@steven-johnson I dont have permissions to push in the main repo. So I've created a separate PR https://github.com/halide/Halide/pull/3471 for the hexagon related changes

**Comment by steven-johnson** [comment]
I'm going to merge this once all the buildbots go green.

**Comment by dsharletg** [comment]
There are some errors on Hexagon: https://buildbot.halide-lang.org/master/#/builders/26/builds/237

One of them looks suspicious (incorrect results), the other looks like something more basic is wrong. I'm guessing something is wrong with the prebuilt, which is super annoying. We really should get the build bots to build those runtimes instead of checking in prebuilts...

**Comment by steven-johnson** [comment]
Just synced this back to master. I think we're probably ready to land this.

**Comment by steven-johnson** [comment]
Bah, missed Dillon's last comment. Let me doublecheck what is going on with Hexagon first.

**Comment by steven-johnson** [comment]
Yeah, the HVX tests are failing: https://buildbot.halide-lang.org/master/#/builders/26/builds/426
```
dlopenbuf failed
undefined PLT symbol halide_malloc (22) ./libhalide_kernels31.soError at /home/halidenightly/build_bot/worker/linux-64-gcc53-testbranch-trunk/halide/test/correctness/math.cpp:160:
Initialization of Hexagon kernels failed
../halide/Makefile:1611: recipe for target 'quiet_correctness_math' failed
```

...do we need to update prebuilts?

**Comment by dsharletg** [comment]
I don't think so, and that error doesn't make sense given it ran a bunch of other tests successfully first. I'm not sure what is going on. Maybe we should split this PR so the Hexagon change is separate and we can debug it at a different pace.

**Comment by steven-johnson** [comment]
Note that the Mac buildbot failure appears to be legit; we're segfaulting in apps/local_laplacian in a way that isn't happening in master. Either there's a bug in this change, or this change is exposing a latent overread/overwrite that was masked by the previous overallocation strategy.

**Comment by steven-johnson** [comment]
The issue here appears to be that there are situations in which we can generate vector code that can 'underread' a buffer (ie before its proper start); this has been masked by the fact that we've been overallocating for alignment purposes, so we always have a bit of slop there. We really need to either quantify the apron that our allocator needs to provide, or to fix the issue so that the allocator doesn't have to deal with this; I've opened https://github.com/halide/Halide/issues/3738 to deal with this.

**Comment by steven-johnson** [comment]
I'm going to close this; https://github.com/halide/Halide/issues/1044 and https://github.com/halide/Halide/issues/3738 make it impossible to land at present (notwithstanding other considerations).

**Pull request closed without merging**